cassovary-collections/src/main/scala/com/twitter/cassovary/collections/CSeq.scala

```diff
@@ -0,0 +1,166 @@
+package com.twitter.cassovary.collections
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import scala.collection.immutable
+import scala.language.experimental.macros
+
+
+/**
+  * CSeq is fast and memory efficient equivalent of Scala's `immutable.IndexedSeq` with minimal
+  * interface that is
+  * 1) generic;
+  * 2) specialized for `Int` and `Long`;
+  * 3) does not use boxing for `Int` and `Long`;
+  * 4) and uses arrays of primitive types underneath.
+  *
+  *
+  * @tparam T type of element
+  */
+trait CSeq[@specialized(Int, Long) T] {
+  def apply(idx: Int): T
+  def length: Int
+
+  /**
+    * @return A sliced view of this CSeq.
+    */
+  def slice(from: Int, until: Int)(implicit csf: CSeqFactory[T]) : CSeq[T] = {
+    val new_until = math.min(length, until)
+    if (from >= new_until) {
+      csf.empty
+    } else {
+      csf.fromCSeq(this, from, new_until)
+    }
+  }
+
+  /**
+    * @return Scala's `Indexed.Seq` that wraps this CSeq. Uses boxing.
+    */
+  def toSeq: CSeqWrapper[T] = new CSeqWrapper[T](this)
+
+  def isEmpty = length == 0
+
+  def foreach[U](f: T => U): Unit = {
+    var i: Int = 0
+    while (i < length) {
+      f(apply(i))
+      i = i + 1
+    }
+  }
+
+  /** Methods below are not optimized */
+  override def hashCode(): Int = toSeq.hashCode()
+
+  override def equals(obj: scala.Any): Boolean = obj match {
+    case cseq: CSeq[T] => toSeq.equals(cseq.toSeq)
+    case _ => false
+  }
+}
+
+class CSeqWrapper[T](underlying: CSeq[T]) extends immutable.IndexedSeq[T] {
+  override def length: Int = underlying.length
+
+  override def apply(idx: Int): T = underlying.apply(idx)
+}
+
+abstract class CSeqFactory[@specialized(Int, Long) T] {
+  def ofDim(n: Int): CSeq[T]
+
+  def fromCSeq(arr: CSeq[T], from: Int, until: Int): CSeq[T]
+
+  def fromArray(arr: Array[T]): CSeq[T]
+
+  def empty: CSeq[T]
+}
+
+/**
+  * Companion object that should be used, when creating `CSeq`s.
```

Comment:
Please change comment to "Companion object to CSeq. This object's apply methods should be used to create new `CSeq`s."